src/index.ts

```diff
@@ -535,6 +535,11 @@ export class ValidationService {
             // we could use 'matches', but that's newer than querySelectorAll so we'll keep it simple and compatible.
             forms.push(root);
         }
+        // If root is the descendant of a form, we want to include that form too.
+        const containingForm = (root as HTMLElement).closest('form');
```

Comment:
Let's actually check the type to make sure `DocumentFragment` keeps working (#45).
```suggestion
        const containingForm = (root instanceof Element) ? root.closest('form') : null;
```

Seems extremely unlikely there would be a `form` in non-HTML, but technically [`closest()` comes from `Element`](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest).

Reply:
I haven't had a chance to look at this closely, but when I see `closest` here, my concern is cases where an input is not inside a form, but declares a form. Do we handle that here?

```html
<html>
<body>
<form id="some-form"></form>
<input name="whatevs" type="input" form="some-form" />
</body>
</html>

Reply:
That is currently being handled separately a few lines above.

```ts
/* If a validation span explicitly declares a form, we group the span with that form. */
let validationMessageElements = Array.from(root.querySelectorAll<HTMLElement>('span[form]'));
for (let span of validationMessageElements) {
    let form = document.getElementById(span.getAttribute('form'));
    if (form) {
        this.pushValidationMessageSpan(form, span);
    }
}
```

Reply:
@LiteracyFanatic that code handles a `span` with a `form` attribute (non-standard HTML). @haacked is asking about an `input` (or `select`/`textarea`) that references a `form` that has not otherwise been discovered as descendent/self/ancestor of the scanned `root`.

I think he's right that scenario isn't covered anywhere, but I'd be content to wait for someone to report that as a bug with a motivating example. At this point we haven't necessarily scanned `root`'s validatable elements, so it's not obvious how we'd identify which additional forms to scan for validation messages. The workaround would be to manually scan the other `form`.

Reply:
Ok, that makes sense. Thanks for the clarification.